pkg/types/contract_reader_cache.go

```diff
@@ -0,0 +1,441 @@
+	params, returnVal any,
+) error {
+	return c.base.GetLatestValue(ctx, readIdentifier, confidenceLevel, params, returnVal)
+}
+
+func (c *CachedContractReader) startSubproc(fn func(), frequency time.Duration) {
+	c.subprocs = append(c.subprocs, make(chan struct{}))
+
+	go func(chClose <-chan struct{}, fn func(), freq time.Duration) {
+		for {
+			timer := time.NewTimer(utils.WithJitter(freq))
+
+			select {
+			case <-timer.C:
+				timer.Stop()
+				fn()
+			case <-chClose:
+				timer.Stop()
+
+				return
+			}
+		}
+	}(c.subprocs[len(c.subprocs)-1], fn, frequency)
+}
+
+func (c *CachedContractReader) clean() {
+	forRemoval := make([]string, 0)
+
+	c.dataLock.RLock()
+
+	for keyStr, data := range c.data {
+		key, keyExists := c.keyLookup[keyStr]
+		if keyExists {
+			if key.staleness == 0 {
+				continue
+			}
+
+			if time.Since(data.LastUpdate()) > key.staleness {
+				forRemoval = append(forRemoval, keyStr)
+			}
+		}
+	}
+
+	c.dataLock.RUnlock()
+
+	c.dataLock.Lock()
+
+	for _, key := range forRemoval {
+		delete(c.data, key)
+		delete(c.keyLookup, key)
+	}
+
+	c.dataLock.Unlock()
+}
+
+func (c *CachedContractReader) pollAndCache() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), pollInterval)
+
+	for _, key := range c.polling {
+		if ctx.Err() != nil {
+			break
+		}
+
+		config, ok := c.getStrategy(key.readIdentifier)
+		if !ok {
+			continue
+		}
+
+		pollConfig := config.(PollAndCache)
+		data := c.mustGetData(key)
+
+		if time.Since(data.LastUpdate()) < pollConfig.Interval {
+			continue
+		}
+
+		_ = c.fallThroughAndCache(ctx, key, pollConfig.Value)
+	}
+
+	cancel()
+}
+
+func (c *CachedContractReader) getStrategy(readIdentifier string) (any, bool) {
+	c.strategyLock.RLock()
+	defer c.strategyLock.RUnlock()
+
+	strategy, ok := c.strategies[readIdentifier]
+
+	return strategy, ok
+}
+
+func (c *CachedContractReader) setStrategy(readIdentifier string, strategy any) {
+	c.strategyLock.Lock()
+	defer c.strategyLock.Unlock()
+
+	c.strategies[readIdentifier] = strategy
+}
+
+func (c *CachedContractReader) mustGetData(key cacheKey) *cachedValue {
+	c.dataLock.RLock()
+	data, hasData := c.data[key.String()]
+	c.dataLock.RUnlock()
+
+	if !hasData {
+		data = &cachedValue{}
+
+		c.dataLock.Lock()
+		c.data[key.String()] = data
+		c.keyLookup[key.String()] = key
+		c.dataLock.Unlock()
+	}
+
+	return data
+}
+
+func newCacheKey(readIdentifier string, confidence primitives.ConfidenceLevel, params any, staleness time.Duration) cacheKey {
+	return cacheKey{
+		readIdentifier: readIdentifier,
+		confidence:     confidence,
+		params:         params,
+		staleness:      staleness,
+	}
+}
+
+type cacheKey struct {
+	readIdentifier string
+	confidence     primitives.ConfidenceLevel
+	params         any
+	staleness      time.Duration
+}
+
+func (k cacheKey) String() string {
+	return fmt.Sprintf("%s::%s::%+v", k.readIdentifier, reflect.TypeOf(k.params), k.params)
```

Comment:
I'm worried about these keys staying aligned with how we interpret these `any` params. Ideally, we could limit ourselves to `comparable` types so that we don't have to serialize at all. However, if we must serialize, then should we use JSON instead, since we already depend on that elsewhere?

Reply:
The cache key is only used locally and is only intended to identify conditions where read parameters match. Is JSON necessary to accomplish this or can we just use Go's value stringification instead?

Reply:
> where read parameters match

How do we know that the way you are printing them is aligned with the user and the chain reader's usage? Using JSON is a trivial way to ensure this isn't a concern, because we already use it to communicate between them.

Furthermore, what if some params contain maps with a randomized iteration order? This breaks the cache key in both cases.

Reply:
Fair point on the JSON.

Key order is also a good point on maps. This would need to be a special case on stringifying the parameters. A special encoder would need to exist that would put the keys in an alphanumeric order in the encoding process.

Reply:
We could use the new json v2 package. You can toggle determinism: https://pkg.go.dev/github.com/go-json-experiment/json#Deterministic
It should be much faster too.

Reply:
I was actually looking at V2 because I remembered it had encoding options. The deterministic option looks like the right solution.